[D2M] Initial implementation of GenericOpBufferAnalysis

bgrady-tt opened the pull request:
### Ticket
Link to Github Issue

### Problem description

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] New/Existing tests provide coverage for changes

vroubtsovTT commented:
Is this going to be an implementation of 

> @vroubtsovTT this is something resembling the BlockFactorAnalysis I discussed in the face-to-face. For now, it basically just allows querying buffer memref shapes for single/double buffer cases.

Is this going to be an implementation of #4801 ?

bgrady-tt commented:
@vroubtsovTT 
> Is this going to be an implementation of #4801 ?

Yes, as-is this amounts to just a query class for stream buffer sizes. 

In the medium term, I see this evolving into a kind of unified performance analysis for different blocking schemes. So the indirect naming reflects that it is intended to not be _just_ for buffer sizing.

vroubtsovTT commented on test/unittests/Allocation/TestBlockFactorAnalysis.cpp:
```diff
@@ -0,0 +1,263 @@
+}
+
+TEST_F(BlockFactorAnalysisTest,
+       CanPerformBlockFactorAnalysis1x4ShardedGenericOp) {
```
Am I correct in reading CanPerformBlockFactorAnalysis1x1ShardedGenericOp, CanPerformBlockFactorAnalysis2x2ShardedGenericOp, and CanPerformBlockFactorAnalysis1x4ShardedGenericOp as being the same unit test but with different parameters?

gtest supports value- and type-parameterized tests for exactly this situation. Over time, I've come to think that type-parameterized are the most universal type. They are a bit painful to set up but once I understood the macro recipe it became cookie cutter. Have a look at TYPED_TEST in TestAllocation.cpp; the doc is https://google.github.io/googletest/reference/testing.html#TYPED_TEST_SUITE

vroubtsovTT commented on include/ttmlir/Dialect/D2M/Analysis/BlockFactorAnalysis.h:
```diff
@@ -0,0 +1,69 @@
+      : constraints(constraints) {}
+  ~BlockFactorAnalysis() = default;
+
+  BlockFactorAnalysis(const BlockFactorAnalysis &) = default;
+  BlockFactorAnalysis &operator=(const BlockFactorAnalysis &) = default;
+  BlockFactorAnalysis(BlockFactorAnalysis &&) = default;
+  BlockFactorAnalysis &operator=(BlockFactorAnalysis &&) = default;
```
Do we expect the caller to want/be able to copy BlockFactorAnalysis around? When in doubt I default such classes to be non-copyable/non-moveable because that's safer and undoing it later is always a non-breaking code change.

vroubtsovTT commented on include/ttmlir/Dialect/D2M/Analysis/BlockFactorAnalysis.h:
```diff
@@ -0,0 +1,69 @@
+    float predictedRuntimeCost = 0.0f;
+  };
+
+  BlockFactorAnalysis() = default;
+  explicit BlockFactorAnalysis(BlockFactorAnalysisConstraints constraints)
+      : constraints(constraints) {}
```
Is default-constructing equivalent to a particular choice of `BlockFactorAnalysisConstraints`? The default constructor should be doc'ed to make that clear.

bgrady-tt replied:
fixed

vroubtsovTT commented on lib/Dialect/D2M/Analysis/BlockFactorAnalysis.cpp:
```diff
@@ -0,0 +1,36 @@
+  SmallVector<SmallVector<int64_t>> operandShardShapes =
+      op.getOperandShardShapes(/*convertTileToScalar=*/false);
+  for (auto [operandIndex, operand] : llvm::enumerate(op.getOperands())) {
+    size_t num_buffers =
```
naming style

bgrady-tt replied:
fixed

vroubtsovTT commented on lib/Dialect/D2M/Analysis/BlockFactorAnalysis.cpp:
```diff
@@ -0,0 +1,36 @@
+    BufferSetting bufferSetting;
+    bufferSetting.bufferShape = operandShardShapes[operandIndex];
+    bufferSetting.numBuffers = num_buffers;
+    result.operandBufferSettings.push_back(bufferSetting);
```
As a code pattern, I'd normally `emplace_back(std::move(bufferSetting))` if it's the last use of `bufferSetting` in this loop.

vwellsTT replied:
I agree with using std::move, but not emplace_back, FWIW.  Milan sent me a stackoverflow link a while ago which changed my views, let me see if I can find.  But the TL;DR is that `push_back` is safer, and therefore better to use for cases where you don't want to use the `emplace_back` construct-in-place feature

bgrady-tt replied:
I will add a `std::move` there, although since its trivially copyable data on the stack I don't think it makes much difference.

I have seen these arguments before against defaulting to `emplace_back`, and so avoid it unless the code path is expected to be hot or constructing in place is desirable for other reasons.

vroubtsovTT commented on test/unittests/Allocation/TestBlockFactorAnalysis.cpp:
```diff
@@ -0,0 +1,263 @@
+  mlir::tt::d2m::BlockFactorAnalysis analysisSingle(singleBufferConstraints);
+  auto bufferConfigsSingle = analysisSingle.analyzeGenericOp(genericOp);
+
+  EXPECT_FALSE(bufferConfigsSingle.empty());
```
like above, this should probably be ASSERT_FALSE, or the next line segfaults

bgrady-tt replied:
fixed.

vroubtsovTT reviewed (commented):
some more comments

vroubtsovTT commented on include/ttmlir/Dialect/D2M/Analysis/BlockFactorAnalysis.h:
```diff
@@ -0,0 +1,69 @@
+  BlockFactorAnalysis &operator=(BlockFactorAnalysis &&) = default;
+
+  // Analyze a ttir::GenericOp and return buffering
+  SmallVector<BufferConfig> analyzeGenericOp(GenericOp op);
```
Ok with your suggestion but would point out that such "caching" in c++ APIs is often done using state fields marked with `mutable`, since the API remains "semantically" const from the caller's point of view.

bgrady-tt replied:
That is a good suggestion; I will make it `const` and insert `mutable` to the cache member var if needed later.

bgrady-tt commented:
@vroubtsovTT I finished addressing all of your comments; I believe in all cases I ended up applying changes (except emplace_back). I also renamed the class to `GenericOpBufferAnalysis`, as I think this is more descriptive of what it does right now.

vroubtsovTT commented on include/ttmlir/Dialect/D2M/Analysis/GenericOpBufferAnalysis.h:
```diff
@@ -0,0 +1,57 @@
+
+  /// Buffer settings for a single operand.
+  struct BufferSetting {
+    SmallVector<int64_t> bufferShape; // in terms of elements
```
For comments like this one that is meant to be an "after member" comment I use `///< ...` syntax, one of the allowed variants, per https://www.doxygen.nl/manual/docblocks.html#memberdoc

vroubtsovTT commented on lib/Dialect/D2M/Analysis/GenericOpBufferAnalysis.cpp:
```diff
@@ -0,0 +1,36 @@
+    result.operandBufferSettings.push_back(std::move(bufferSetting));
+  }
+
+  results.push_back(result);
```
> I believe in all cases I ended up applying changes (except emplace_back).

nit: this is one more case here where you could be pushing std::move()d `result`, analogous reason to `bufferSetting`.

vwellsTT replied:
heck, the most optimal thing to do would be to use `emplace_back` with proper fields to construct the `OpConfig` truly in place here 🙂

vroubtsovTT commented on test/unittests/Allocation/TestGenericOpBufferAnalysis.cpp:
```diff
@@ -0,0 +1,218 @@
+            params.expectedBufferShape);
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    GenericOpBufferAnalysisTests, GenericOpBufferAnalysisParamTest,
+    ::testing::Values(
+        GenericOpBufferAnalysisTestParams{{1, 1}, {2, 4}, {2, 4}, "1x1Sharded"},
+        GenericOpBufferAnalysisTestParams{{2, 2}, {4, 8}, {2, 4}, "2x2Sharded"},
+        GenericOpBufferAnalysisTestParams{
+            {1, 4}, {2, 8}, {2, 2}, "1x4Sharded"}));
+
+TEST_F(GenericOpBufferAnalysisTest, CanAnalyzeBufferingStrategies) {
```
nice!

bgrady-tt commented on lib/Dialect/D2M/Analysis/GenericOpBufferAnalysis.cpp:
```diff
@@ -0,0 +1,36 @@
+  SmallVector<SmallVector<int64_t>> operandShardShapes =
+      op.getOperandShardShapes(/*convertTileToScalar=*/false);
+  for (auto [operandIndex, operand] : llvm::enumerate(op.getOperands())) {
+    size_t numBuffers = (constraints.bufferingStrategy ==
```
fixed!
